File: course_project/src/main/java/ru/ssau/course_project/service/ProjectService.java
package ru.ssau.course_project.service;

import jakarta.persistence.EntityNotFoundException;


import ru.ssau.course_project.entity.dto.ProjectDto;

import java.time.LocalDate;
import java.util.List;

public interface ProjectService {

    ProjectDto create(ProjectDto dto) throws EntityNotFoundException, IllegalArgumentException;

    ProjectDto update(ProjectDto dto) throws EntityNotFoundException;

    void delete(Long id);

    List<ProjectDto> findAll();

    ProjectDto findById(Long id) throws EntityNotFoundException;

    List<ProjectDto> findAllByStatus_NameIgnoreCase(String status);

    List<ProjectDto> findAllByStartDateAfter(LocalDate startDate);

    List<ProjectDto> findAllByEmployeeId(Long id) throws EntityNotFoundException;

    List<ProjectDto> findMyProjects();

    void addToTeam(Long projectId, List<Long> employeeIds) throws EntityNotFoundException;
    void removeFromTeam(Long projectId, Long employeeId) throws EntityNotFoundException;

}

File: course_project/frontend/src/app/services/project.service.ts
import {inject, Injectable} from '@angular/core';
import {Project} from '../models/project';
import {Observable} from 'rxjs';
import {HttpClient, HttpParams} from '@angular/common/http';
import {environment} from '../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class ProjectService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/project`;
  private myProjectsUrl = `${environment.apiUrl}/my-projects`;


  create(project: Project): Observable<Project> {
        return this.http.post<Project>(this.baseUrl, project);
  }

  update(project: Project): Observable<Project> {
    return this.http.put<Project>(this.baseUrl, project);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  findById(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.baseUrl}/${id}`);
  }
  findAll(): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.baseUrl}/all`);
  }

  findAllByStatus(status: string): Observable<Project[]> {
    let params = new HttpParams().set('status', status);
    return this.http.get<Project[]>(`${this.baseUrl}/all/by-status`, { params });
  }

  findAllByStartDateAfter(afterDate: string): Observable<Project[]> {
    let params = new HttpParams().set('after', afterDate);
    return this.http.get<Project[]>(`${this.baseUrl}/all/by-start-date`, { params });
  }

  findAllByEmployeeId(employeeId: number): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.baseUrl}/all/${employeeId}`);
  }

  findMyProjects(status?: string): Observable<Project[]> {
    let params = new HttpParams();

    if (status) {
      params = params.set('status', status);
    }
    return this.http.get<Project[]>(this.myProjectsUrl, { params });
  }


  findBySprintId(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.myProjectsUrl}/by-sprint-id/${id}`)
  }

  addToTeam(projectId: number, employeeIds: number[]): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${projectId}/add-to-team`, employeeIds);
  }

  removeFromTeam(projectId: number, employeeId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${projectId}/delete-from-team/${employeeId}`, null);
  }


}
